### ISSUE TITLE
Behaviour of layer-locking double tap

### ISSUE THREAD

**Issue by kareltucek** [UltimateHackingKeyboard/agent#749]
Hello,

when discussing timeouts for modifier locking, I have noticed one thing - layer does not get locked when two different hardware keys with the same layer modifier get tapped (I had tried locking that way, because with the current timeout, the only way to actually lock a layer is by a well-synchronized tap by two different fingers).

My argumentation and ideas regarding this go as follows (let's assume we are talking about fn layer with default factory keymap, w.l.o.g.):
- Locking by doubletap of the same key is "slow" and therefore leads to accidental locking (as reported by other users)
- Layers could be locked by simultaneous or almost simultaneous tapping of different (w.r.t. hardware) fn keys. This would allow comfortable locking without the danger of accidental locking, since timeout needed for "parallel" doubletap is much lower than for "serial". 
- Technically taken, double tap performed on the same "virtual" key is still double tap, even when performed via two different hardware keys. This leads me to a conclusion that the current behaviour might even be considered wrong.
- I thought that his might be easy to set up via binding fns to 'layer lock' on fn layer. However, layer dialog does not allow initiating layer lock from non-default layer.

I would like to ask your opinion on whether the current behaviour is correct behaviour or not, and whether it is the desirable behaviour or not (and why).

**Comment by mondalaci** [comment]
Hi Karel,

The current behavior is considered to be correct. I think that double-tapping a single layer switcher key is far more intuitive than double-tapping with multiple layer switcher keys of the same type.

I'm surprised that you can't lock a layer with a double-tap. I admit that the current hardcoded timeout is rather low, but people don't seem have trouble with this.

If we made the double-tapping of different keys of the same layer switcher key type lock the relevant layer, it'd ask for different double-tap timeout options for layers that have multiple layer switcher keys and layers that have a single layer switcher key (typically Mouse) to minimize the chance of accidental locking, resulting in more user-facing complexity in Agent. This doesn't seem right because based on the feedback we receive, nobody else has requested this feature, yet it'd complicate configuration.

In the future, a global double-tap timeout and per-key "double-tap to layer lock" flags will be configurable, and by default layer locking will only be enabled for the Mouse key. I think this will work fine for the vast majority of users.

I'm closing this issue, but feel free to follow up.

**Comment by epilgrim** [comment]
I'll chime in just to say that I also thought that the layer locking was broken, until I read your comment of doing it faster. I can never lock it with normal typing. It is mostly fine, because i rarely lock layers... but when I want to do so I have to move my hand specifically for this and hit the keys as fast as I can

**Comment by mondalaci** [comment]
Thanks for chiming in @epilgrim.

Which layers do you guys typically lock, and how frequently?

**Comment by mondalaci** [comment]
According to my knowledge, the vast majority of users largely sticks to the default keymap. If so, from a practical standpoint, it makes a lot of sense to lock Mouse because the pinky gets tired. It makes some sense to lock Mod to navigate without holding, but this is far less frequently used than Mouse. As for Fn, it doesn't make much practical sense to lock media keys.

I think given the above, the new features to be implemented and new default configuration should work pretty well. Some may want to also configure the right Mod case button, but not the left Mod key to lock on double-tap. I think it's a reasonable configuration for those who want to lock Mod.

Marking this as an exotic feature as suggested.

**Comment by flyinprogrammer** [comment]
@mondalaci 

> I'm surprised that you can't lock a layer with a double-tap.

Hopefully this provides some good user feedback: https://youtu.be/wOfdiM2mLqg
(My apologies for the horizontal filming, I am an idiot.)

After about 5 more minute she was able to start getting it to lock with reasonable accuracy, I'm still struggling lol. I have the Red caps, and it seems the trick is to not let the key come all the way back up before the second tap, which takes some training for my apparently weak, and slow pinky. 

(Also I cannot express enough how awesome this product is, and how happy I am with this tool! It's truly exceptional. This feedback simply comes from a place of respect and love.)

**Comment by mondalaci** [comment]
@flyinprogrammer Thanks so much for the nice words and video! Which firmware version are you using?

**Comment by flyinprogrammer** [comment]
Firmware: `8.4.0`
with Agent `1.2.7` 
on OS X `10.13.6`

**Comment by mondalaci** [comment]
@flyinprogrammer Your observation turned out to be very useful. I realized that we fixed the timer in 8.3.3 which made it tick twice as fast as before. As a side-effect, it also made the double tap to lock timeout half as long as before. This is why it was so hard to trigger it in firmware versions 8.3.3 - 8.4.2. Just released 8.4.3 which compensates for the timer fix, and the timeout should be easy to trigger again, the same as before 8.3.3.
